### ISSUE TITLE
Race conditions in lifecycle

### ISSUE THREAD

**Issue by kriskowal** [yarpc/yarpc-go#709]
There are race conditions in lifecycle that occasionally manifest with the stress test:

```
--- FAIL: TestLifecycleOnce (0.21s)
    --- FAIL: TestLifecycleOnce/Stress_test (0.08s)
        --- FAIL: TestLifecycleOnce/Stress_test/action_#0:_sync.ConcurrentAction (0.08s)
        Error Trace:    lifecycle_action_test.go:55
        Error:      	Not equal:
                    	expected: 2
                    	received: 3
```

The cause is that we’re using an enum for state, but the atomic integer does not guarantee monotonicity. We should revisit the design using a bank of boolean atomics to indicate which transitions have occurred and use CAS to advance them monotonically.

```
Start()
   if starting.CAS(true) {
      <-startedCh
   } else {
       doStart()
       started.Store(true)
       close(startedCh)
   }
Stop()
   if starting.CAS(true) {
     <-startedCh
   } else {
       started.Store(true)
       close(startedCh)
   }
   if stopping.CAS(true) {
       <-stoppedCh
   } else {
        doStop()
        stopped.Store(true)
        close(stoppedCh)
   }
IsStarting()
    return starting.Load() && !started.Load()
IsRunning() # Available / Started
    return started.Load() && !stopping.Load()
IsStopping()
    return stopping.Load() && !stopped.Load()
IsStopped()
    return stopped.Load()
```

**Comment by kriskowal** [comment]
It might be fine just to adjust the test to accept >= expected state, to verify monotonicity.

**Comment by kriskowal** [comment]
cc @willhug for ¢¢.

**Comment by willhug** [comment]
The error here is a bit simpler, and the solution provided would not actually make the test any less flaky.  The issue is that there is a race in the test infrastructure itself (not the LifecycleOnce).

In lifecycle_action_test we run
```
	err := l.Start(func() error {
		if a.Wait > 0 {
			time.Sleep(a.Wait)
		}
		return a.Err
	})
	assert.Equal(t, a.ExpectedErr, err)
	assert.Equal(t, a.ExpectedState, LifecycleState(l.state.Load()))
```
the asserts here are not done inside a lock of any kind, which means that there are no guarantees of order combined with the `ConcurrentAction` we use.  So in the concurrent test:

```
StartAction{
	ExpectedState: Running, // Fine
},
StartAction{
	ExpectedState: Running, // WHOOPS, a Stop happened after the action finished, but before we asserted on the result, the state is now "Stopping" instead.
},
StopAction{
	ExpectedState: Stopped,
},
```
We might consider hardening the test infrastructure to have a lock between Actions, so only one can run at a time, but that would sort of defeat the point of the concurrent test.  Instead we should probably just remove the assertions on lifecycle state when we are doing concurrent actions since actions are not atomically locked.

**Comment by kriskowal** [comment]
We can just relax the ExpectedState assertion to verify that the state is == or >.

Are you confident that Start() and Stop() block until the Running and Stopped states respectively, even when they do not initiate the transition?

**Comment by akshayjshah** [comment]
Was this resolved in #848?

**Comment by kriskowal** [comment]
That’s the trick!
